src/vulkan/engine_vulkan_debugger.cc

```diff
@@ -0,0 +1,976 @@
+// convenient interface for talking to the debugger. Client also provides basic
+// immutable data caching to help performance.
+class Client {
+  static constexpr const char* kLocals = "locals";
+  static constexpr const char* kLane = "Lane";
+
+ public:
+  using ErrorHandler = std::function<void(const std::string&)>;
+  using SourceLines = std::vector<std::string>;
+
+  Client(const std::shared_ptr<dap::Session>& session,
+         const ErrorHandler& onerror)
+      : session_(session), onerror_(onerror) {}
+
+  // TopStackFrame retrieves the frame at the top of the thread's call stack.
+  // Returns true on success, false on error.
+  bool TopStackFrame(dap::integer threadId, dap::StackFrame* frame) {
+    dap::StackTraceRequest request;
+    request.threadId = threadId;
+    auto response = session_->send(request).get();
+    if (response.error) {
+      onerror_(response.error.message);
+      return false;
+    }
+    if (response.response.stackFrames.size() == 0) {
+      onerror_("Stack frame is empty");
+      return false;
+    }
+    *frame = response.response.stackFrames.front();
+    return true;
+  }
+
+  // FrameLocation retrieves the current frame source location, and optional
+  // source line text.
+  // Returns true on success, false on error.
+  bool FrameLocation(const dap::StackFrame& frame,
+                     amber::debug::Location* location,
+                     std::string* line = nullptr) {
+    location->line = frame.line;
+
+    if (!frame.source.has_value()) {
+      onerror_("Stack frame with name '" + frame.name + "' has no source");
+      return false;
+    } else if (frame.source->path.has_value()) {
+      location->file = frame.source.value().path.value();
+    } else if (frame.source->name.has_value()) {
+      location->file = frame.source.value().name.value();
+    } else {
+      onerror_("Frame source had no path or name");
+      return false;
+    }
+
+    if (location->line < 1) {
+      onerror_("Line location is " + std::to_string(location->line));
+      return false;
+    }
+
+    if (line != nullptr) {
+      SourceLines lines;
+      if (!SourceContent(frame.source.value(), &lines)) {
+        return false;
+      }
+      if (location->line > lines.size()) {
+        onerror_("Line " + std::to_string(location->line) +
+                 " is greater than the number of lines in the source file (" +
+                 std::to_string(lines.size()) + ")");
+      }
+      *line = lines[location->line - 1];
+    }
+
+    return true;
+  }
+
+  // SourceContext retrieves the the SourceLines for the given source.
+  // Returns true on success, false on error.
+  bool SourceContent(const dap::Source& source, SourceLines* out) {
+    auto path = source.path.value("");
+    if (path != "") {
+      auto it = sourceCache_.by_path.find(path);
+      if (it != sourceCache_.by_path.end()) {
+        *out = it->second;
+        return true;
+      }
+
+      std::ifstream file(path);
```

Comment:
This should, probably, call back to the embedder to load the file. The amber library doesn't know how to read files anywhere else, it assumes the embedder will handle that for it.

Reply:
By "embedder", you mean something like dEQP?
Acknowledged for now - I'll have a think about this.

Reply:
Yea, even the `amber` sample app is an 'embeddeder' of the amber library. Adding a `TODO` here is fine and file a bug and we can come back to it later.

Reply:
Done: https://github.com/google/amber/issues/777